syntax="proto3";

package transportCatalogueSerialized;

import "transport_router.proto";

message Stop {
    uint32 unique_index = 1;
    string name = 2;
    double latitude = 3;
    double longitude = 4;
}

message Bus {
    uint32 unique_index = 1;
    string name = 2;
    repeated uint32 stops = 3;
    bool roundtrip = 4;
}

message Distance {
    uint32 start_stop = 1;
    uint32 finish_stop = 2;
    uint32 distance = 3;
}

message Rgb {
    uint32 red = 1;
    uint32 green = 2;
    uint32 blue = 3;
}

message Rgba {
    uint32 red = 1;
    uint32 green = 2;
    uint32 blue = 3;
    double opacity = 4;
}

enum ColorTypes {
    MONOSTATE = 0;
    STRING = 1;
    RGB = 2;
    RGBA = 3;
}

message RenderSettings {
    double width = 1;
    double height = 2;
    double padding = 3;
    double line_width = 4;
    double stop_radius = 5;
    uint32 bus_label_font_size = 6;
    double bus_label_offset_first = 7;
    double bus_label_offset_second = 8;
    uint32 stop_label_font_size = 9;
    double stop_label_offset_first = 10;
    double stop_label_offset_second = 11;
    double underlayer_width = 12;

    // Тип показывает, в какой из трех очередей искать следующий цвет
    // Первое число - тип для underlayer_color, все последующие для color_palette
    repeated ColorTypes color_types = 13;

    repeated string color_palette_string = 14;
    repeated Rgb color_palette_rgb = 15;
    repeated Rgba color_palette_rgba = 16;

}

message TransportCatalogue {
    repeated Stop stops = 1;
    repeated Bus buses = 2;
    repeated Distance distances = 3;
    RenderSettings render_settings = 4;
    TransportRouter transport_router = 5;
}